spec/models/search_spec.rb

```diff
@@ -0,0 +1,297 @@
+
+        it "returns the record" do
+          expect(search).to eq([prepare.id])
+        end
+      end
+
+      context "with non-standard magistrate application" do
+        let(:prepare) { build(:submission, :with_nsm_version).tap(&:save) }
+
+        it "returns the record" do
+          expect(search).to eq([prepare.id])
+        end
+      end
+    end
+
+    context "when it does not match all parts of defendant name" do
+      let(:query) { "bob bloggs" }
+
+      context "with with prior authority application" do
+        let(:prepare) { build(:submission, :with_pa_version).tap(&:save) }
+
+        it "returns the record" do
+          expect(search).to be_empty
+        end
+      end
+
+      context "with non-standard magistrate application" do
+        let(:prepare) { build(:submission, :with_nsm_version).tap(&:save) }
+
+        it "returns the record" do
+          expect(search).to be_empty
+        end
+      end
+    end
+
+    context "when search text is partial defendant name" do
+      let(:query) { "joe" }
+
+      context "with prior authority application" do
+        let(:prepare) { build(:submission, :with_pa_version).tap(&:save) }
+
+        it "returns the record" do
+          expect(search).to eq([prepare.id])
+        end
+      end
+
+      context "with non-standard magistrate application" do
+        let(:prepare) { build(:submission, :with_nsm_version).tap(&:save) }
+
+        it "returns the record" do
+          expect(search).to eq([prepare.id])
+        end
+      end
+    end
+
+    context "when user name is Jason/Jim" do
+      let(:query) { "Jason/Jim" }
+
+      context "with prior authority application" do
+        let(:prepare) { build(:submission, :with_pa_version, defendant_name: "Jason/Jim Read").tap(&:save) }
+
+        it "returns the record" do
+          expect(search).to eq([prepare.id])
+        end
+
+        context "with additional partial name matches" do
+          let(:prepare) do
+            [
+              build(:submission, :with_pa_version, defendant_name: "Jason/Jim Read").tap(&:save),
+              build(:submission, :with_pa_version, defendant_name: "Jason Write").tap(&:save),
+              build(:submission, :with_pa_version, defendant_name: "Jim Type").tap(&:save),
+              build(:submission, :with_pa_version, defendant_name: "Jack Burns").tap(&:save),
+            ]
+          end
+
+          it "only matches when both name parts are present" do
+            expect(search).to eq([prepare[0].id])
+          end
+        end
+
+        context "when searching on the first part of first name only" do
+          let(:query) { "Jason" }
+
+          it "returns the record" do
+            expect(search).to eq([prepare.id])
+          end
+        end
+
+        context "when searching on the last part of first name only" do
+          let(:query) { "Jim" }
+
+          it "returns the record" do
+            skip "TBC: not surrently working due to `/`"
```

Comment:
one of the remaining issue that I am aware of.

could be solved by replacing `/` with `-` but not 100% sure if this is a good idea.